### DISCUSSION TITLE
Compiler Optimizations

### DISCUSSION THREAD

**Discussion by black7375** [Floorp-Projects/Floorp#421] [Feature Request]
I recently knew the [Mercury browser](https://thorium.rocks/mercury), but I think it's quite interesting.
It is known to use various compiler [optimization strategies](https://thorium.rocks/optimizations) such as AVX, AES, -O3, Polly, ThinLTO and PGO.

This project and Firefox([v115](https://www.mozilla.org/en-US/firefox/115.0/system-requirements/), [v116](https://www.mozilla.org/en-US/firefox/116.0.1/system-requirements/)) dropped support Windows7/8, so I think it's a good time to look at it.

Here is a repo
- https://github.com/Alex313031/Mercury/tree/main

**Comment by Ryosuke-Asano** [comment]
Mercury is interesting. However, we have observed that in some cases the compiler flags used are backfiring, and we do not know what optimization will result in the fastest configuration, as O3 optimization is fastest on Linux and Mac, but worse on Windows.


![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/203b6558-6711-41a8-a4d3-946099031914)

It's a fine line, as some benchmarks show faster results with Floorp. I don't mean to denigrate them, and I want to help, so I make suggestions on GitHub and so on. https://github.com/Alex313031/Mercury/issues/9

![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/598c60fc-5f4c-4b7d-bc7e-6606ba02905d)
![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/75177ff7-d326-4644-8ca4-0aa67f0b5295)

Floorp was the first to introduce PGO and LTO in a Firefox derivative (Floorp 10.5.0), and Floorp was the first to introduce PGO on the Mac (Floorp 11). We have never compromised on performance.

The conclusion is that we are in a state of wanting to wait and see.

**Comment by Ryosuke-Asano** [comment]
On Mac:

![image](https://github.com/Floorp-Projects/Floorp/assets/73892113/fbae796a-9b0d-43fc-9a14-7c9a0b240216)

**Comment by black7375** [comment]
> as O3 optimization is fastest on Linux and Mac, but worse on Windows

Yes, so the performance should always be measured, and Waterfox seems to have different optimizations for each OS.
I have been a long time ago, I benchmarked [tete](http://www1.plala.or.jp/tete009/en-US/software.html), [pcxFirefox](https://sourceforge.net/projects/pcxfirefox/), [lawlietfox](https://sourceforge.net/projects/lawlietfox/) or others, but I remember that the performance regression sometimes appeared as unexpected.
- https://github.com/WaterfoxCo/Waterfox/commit/ab1549b21ff3e632dff60bebad8b5d57ec774fc6

This is a list of benchmarks provided by Mozilla.
I think it's automated, but I don't know exactly what code is.
- https://arewefastyet.com/

**Comment by Ryosuke-Asano** [comment]
Yes, I agree.

Optimization in derivation is of course important, but the most important thing is the improvement of the Gecko engine and the existence of disadvantages cannot be discarded now, so compiler optimization is meaningless now in my opinion because of demerit exist.

Waterfox introduced multifaceted optimization in G6.0b2, but PGO is so powerful that it is honestly hazy. I had also made major modifications to the settings, but after all, if Mozilla introduces it, it will ruin its value as a derivative, so I'm not sure about it.

Mozilla itself is also making daily efforts to improve performance.

**Comment by gunir** [comment]
Recently Waterfox introduced `polyhedral optimisation`: https://github.com/WaterfoxCo/Waterfox/releases/tag/G6.0b1#fn-3

Should we give it a try sometimes ? It sounds pretty good:

```
As previously mentioned, when you use a web browser, it has to repeat similar tasks over and over - like rendering webpage elements, running JavaScript code, scrolling the page, and displaying videos.

These tasks involve loops, where the browser does the same kind of work over and over. Polyhedral optimisation makes the browser's loops run faster by reorganizing them. It finds the best way to access data repeatedly in the most efficient order.

For example, when scrolling a page with lots of images, polyhedral optimisation could optimise the loops that decode and display the image data. It finds the fastest way to reuse image data in the CPU cache.

This means scrolling feels much smoother. The images decode quicker as you scroll up and down. The browser feels snappier.

So in simple terms, polyhedral optimisation tunes the browser's inner loops to make maximum use of the CPU and memory. This speeds up repetitive tasks like rendering, scrolling, and video playback that rely on loops. The end result is a faster, smoother browsing experience for the user.
```

**Reply by Ryosuke-Asano** [reply]
In fact, Previous my posted screenshot is NOT G5. It is G6.

The optimization is not affect to result.

PGO is important but, oployhedral is not affect...
